### ISSUE TITLE
Proposal to fix model.trainable

### ISSUE THREAD

**Issue by isaacgerg** [fchollet/keras#4510]
I want to propose a change to keras but would like advice on how best to do this change so that PR gets accepted.  I would like to make model.trainable=True flow down so all the layers inside this model are non-trainable. I see a few ways to do this and not sure what the best way is:

1,. Make Model.trainable its own parameter not inherited from Container.trainable.  When it is set to False (or True), set all layers trainable to False.  This will have to run recursively if there are models inside the model.

2. Somehow work with Containers.trainable and when Model.trainable is set, containers somehow knons to update the trainable for all the associated layers in the model.

3. When returning trainable weights, look at model.trainable to override the layer.trainable property.

4. Something else.

Advice greatly appreciated.

Isaac

**Comment by farizrahman4u** [comment]
Why would you want to recursively set trainable for child layers? That would be unexpected behavior. Note that a layer could be part of multiple models, models with different trainable values.

**Comment by isaacgerg** [comment]
Here is a usecase.  Suppose I want to do transfer learning and start with
vggnet and build a model around it.  I would expect vggnet.trainable=False
to disable training of the vggnet, but it doesnt.

    x = AveragePooling2D(pool_size=(2,2))(input)
    x = Flatten()(x)
    x = keras.layers.RepeatVector(3)(x)
    x = Reshape((200, 200, 3))(x)
    vgg1 = keras.applications.vgg16.VGG16(include_top=False)
    vgg1.trainable = False    # Doesnt work
    x = vgg1(x)
    m = keras.layers.Convolution2D(16,3,3, border_mode='same')(x)
    m = keras.layers.advanced_activations.LeakyReLU(alpha=0.01)(m)
    m = keras.layers.SpatialDropout2D(0.5)(m)
    m = keras.layers.Dense(2)(m)

**Comment by isaacgerg** [comment]
Here is a usecase.  Suppose I want to do transfer learning and start with vggnet and build a model around it.  I would expect vggnet.trainable=False to disable training of the vggnet, but it doesnt.

```
    x = AveragePooling2D(pool_size=(2,2))(input)
    x = Flatten()(x)
    x = keras.layers.RepeatVector(3)(x)
    x = Reshape((200, 200, 3))(x)
    vgg1 = keras.applications.vgg16.VGG16(include_top=False)
    vgg1.trainable = False    # Doesnt work
    x = vgg1(x)
    m = keras.layers.Convolution2D(16,3,3, border_mode='same')(x)
    m = keras.layers.advanced_activations.LeakyReLU(alpha=0.01)(m)
    m = keras.layers.SpatialDropout2D(0.5)(m)
    m = keras.layers.Dense(2)(m)
```

@farizrahman4u What else would be reasonable for vgg1.trainable=False to represent?

**Comment by farizrahman4u** [comment]
Update Keras. And how are you checking if it works or not?

**Comment by isaacgerg** [comment]
@farizrahman4u Can you read the original post again? I am looking for suggestions on which way to go to address the issue.

**Comment by farizrahman4u** [comment]
We can't do that. We can't set trainable=False for layers automatically when we set model.trainable=False.
And there is no use for it. If you have a specific issue, post a simplified script that can reproduce it.

**Comment by isaacgerg** [comment]
@farizrahman4u I dont understand.  What does model.trainable=False currently do for a model?

**Comment by farizrahman4u** [comment]
It freezes the model. Weights belonging to that model will not be changed during training.

**Comment by isaacgerg** [comment]
@farizrahman4u It does not freeze the model.  .trainable=False only works on layers, not models. Run the code I posted.

**Comment by farizrahman4u** [comment]
It does work on models and layers. Again, how are you checking if it freezes or not?

**Comment by isaacgerg** [comment]
@farizrahman4u I did a .summary() on my model and the number of parameters did not change.  However, when I manually set all the layers inside of vgg1 to trainable=false, the number of parameters dropped significantly as expected.

When you run my code, how many parameters does vgg1.summary() give you?

Here is a simpler example....

```
vgg = keras.applications.vgg16.VGG16(include_top=False)
vgg.summary()
vgg.trainable=False
vgg.summary()
```

Both summaries report the same number of parameters.  Can you verify on your machine?

**Comment by farizrahman4u** [comment]
Thats a bug in .summary(). Setting trainable=False for layers in a model automatically is not the right way to fix it. Nevertheless the vgg weights will not be trained if you set vgg.trainable=False

**Comment by isaacgerg** [comment]
@farizrahman4u This is helpful to know.  I will write this up as a bug for .summary()

**Comment by farizrahman4u** [comment]
#4513

**Comment by isaacgerg** [comment]
#4514

**Comment by xiaojidan** [comment]
@farizrahman4u  How do you update keras?

**Comment by farizrahman4u** [comment]
`git pull`, followed by `python setup.py install`

**Comment by xiaojidan** [comment]
@farizrahman4u  Thanks a lot.

**Comment by RohitKeshari** [comment]
In one particular application, I need to train only bias of the convolutional operation. Therefore I removed W parameter from trainable_weight. that look like this:  self.trainable_weights = [ self.b]. I save model at 0 epoch and after 200 epoch and I found that W is some how able to learn. Do not know what going on.When I saw model.summary() it showing correct number of learning parameters. Can any one tell me what is wrong here?
